### PR TITLE
feat(construction): add abridged construction and constructionset

### PR DESCRIPTION
**Pull request by tanushree04**
- constructionabridged.py
- constructionsetabridged.py
- materials.py
- Added layers property to construction.py

### CODE REVIEW

File: app/models/energy/construction.py [hunk 1]
```diff
@@ -832,6 +23,9 @@ class EnergyConstructionWindow(BaseModel):
         max_length=100
     )
 
+    layers: List[constr(regex=r'^[\s.A-Za-z0-9_-]*$',
+                        min_length=1, max_length=100)]
+
     materials: List[
```
Comment (macumber) [inline app/models/energy/construction.py:59]: Could we just make materials optional?  Then the "abridged version" is just a construction without that optional field and an "extended version" is a construction with the optional field. There could be a convention that if materials are present, layers must refer to materials nested within the object.

File: app/models/samples/energyconstructionsetabridged.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+construction_set = {
+    'type': 'ConstructionSetAbridged',
+    'name': 'Construction Set',
+    'wall_set': {
+        'type': 'WallSetAbridged',
```
Comment (macumber) [inline app/models/samples/energyconstructionsetabridged.py:5]: Can a ConstructionSetAbridged contain a non-abridged WallSet?

**Review by macumber — COMMENTED** [review]
@tanushree04 @chriswmackey it seems to be that the difference between the abridged and non-abridged objects is the presence of some extra fields in the non-abridged object.  I would consider two approaches:

1) see if we can combine the two objects and just make those fields optional.  That way the abridged and non-abridged objects aren't different types, they are the same type with differences in which optional fields they have. 
2) derive the non-abridged object from the abridged object so it inherits fields from the abridged object.  That way the two objects can share the same schema definitions for the fields they have in common.

Both of these approaches would address the substitutability of an abridged object for a non-abridged one and vice versa, right now that is a little unclear to me.  I do think it would be better to cut down on the number of types so I prefer option 1.  But if you feel that the addition of the extra data justifies a new type then I would suggest naming the abridged object like "Construction" and the object which extends it to add fields (e.g. materials) like "ConstructionExtended".

**Comment by chriswmackey** [comment]
@macumber ,

Thank you for your suggestion but we are positive that these need to be two separate types of schemas.  While you are correct that, in this case of Construction, the Abridged version is just the like full version but lacking a "materials" tag, the situation will be different for other Abridged objects.  For example an ZoneAbridged object simply possesses the name of the ConstructionSet assigned to it in place of full description of the ConstrucionSet. Another example is that a FaceAbridged object that has a custom construction assigned to it will only possess the name of that construction and not a detailed description of it.

While I am not completely against the idea of using inheritance in this case of Constructions, you can see from the example above that this inherit-ability is not something that we can easily replicate across the other objects so it is perhaps not the best idea to set the precedent here.  This is also the only case where we have decided that we will currently support a non-Abridged object so we can just leave it the way it is now and cross the bridge later (we don't need to dwell on it now).

We are also pretty settled on the naming convention with the use of the word "Abridged".  The word "Extended" implies that the object possesses properties beyond what is needed to re-create it and, above all, it is important that users be aware that Abridged objects do not possess everything needed to re-create them and must be a part of a Model to be valid.

File: app/models/energy/construction.py (hunk 1 above)
Comment (chriswmackey) [reply app/models/energy/construction.py:59]: Materials are required to recreate a object and they need to be required on the non-abridged version of the object.  See my reply below for a longer description.

File: app/models/samples/energyconstructionsetabridged.py (hunk 2 above)
Comment (chriswmackey) [reply app/models/samples/energyconstructionsetabridged.py:5]: No, it cannot.  The Abridged schema and the non-Abridged schema are two separate worlds of objects.  The only place where their paths cross are for objects that do not possess other nested objects (ie. materials)

File: app/models/energy/constructionabridged.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+"""Construction Abridged Schema"""
+from pydantic import BaseModel, Schema, validator, ValidationError, constr
+from typing import List, Union
+from enum import Enum
+from uuid import UUID, uuid4
+from datetime import datetime
+
+
+class EnergyConstructionWindowAbridged(BaseModel):
```
Comment (chriswmackey) [inline app/models/energy/constructionabridged.py]: I just realized that I have been naming all construction objects  in the Python code without the use of the word "Energy" and I didn't realize it was different here.
With EnergyMaterials, I realize we wanted to include the word "Energy" to make it clear that we were talking about a different type of material than a radiance material.  However, we don't have the same restriction with Constructions. 

Given this, can we change the names to the following:
`OpaqueConstruction`, `OpaqueConstructionAbridged`, `WindowConstruction`, `WindowConstructionAbridged`.

Hopefully, it should just be a quick Find and Replace, @tanushree04

**Review by chriswmackey — APPROVED** [review]
@tanushree04 ,

The PR looks good to me and thank you for getting it together.  The only comment that I have there is that I realized the naming convention that you have been using for Construction objects is different than the one I have been using and it would be good if we can change the convention used here sooner than later.  I'm approving the PR now and I leave it up to you if you would rather make the naming change within this PR or merge this one in and send another.

**Comment by mostaphaRoudsari** [comment]
I think @macumber's suggestion for inheritance makes sense here. We basically only need to overwrite material field to change the type to str. @chriswmackey why do you think inheritance won't work?

PS: To reply to your comment above. I think we should use inheritance when it works and if there is a case that doesn't work that way they you don't use it not the other way around.

```python
class OpaqueConstructionAbridged(BaseModel):
    """
    Group of objects to describe the physical properties and configuration for
    the building envelope and interior elements that is the walls, roofs, floors,
    and doors of the building.
    """
    type: Enum('OpaqueConstructionAbridged', {
               'type': 'OpaqueConstructionAbridged'})

    name: str = Schema(
        ...,
        regex=r'^[\s.A-Za-z0-9_-]*$',
        min_length=1,
        max_length=100
    )

    layers: List[constr(regex=r'^[\s.A-Za-z0-9_-]*$', min_length=1, max_length=100)] = Schema(
        ...,
        description='List of materials. The order of the materials is from outside to'
        ' inside.',
        minItems=1,
        maxItems=10
    )

    @validator('layers', whole=True)
    def check_num_items(cls, layers):
        "Ensure length of material is at least 1 and not more than 10."
        if len(layers) == 0:
            raise ValidationError(
                'Opaque construction should at least have one material.'
            )
        elif len(layers) > 10:
            raise ValidationError(
                'Opaque construction cannot have more than 10 materials.'
            )
        else:
            return layers


class OpaqueConstruction(OpaqueConstructionAbridged):
    """
    Group of objects to describe the physical properties and configuration for
    the building envelope and interior elements that is the walls, roofs, floors,
    and doors of the building.
    """
    type: Enum('OpaqueConstruction', {
               'type': 'OpaqueConstruction'})

    materials: List[
        Union[
            EnergyMaterial, EnergyMaterialNoMass
        ]
    ] = Schema(
        ...,
        description='List of materials. The order of the materials is from outside to'
        ' inside.',
        minItems=1,
        maxItems=10
    )
```

**Comment by chriswmackey** [comment]
@mostaphaRoudsari ,

Makes sense and just to clarify, I said "I am NOT against the idea of using inheritance in this case of Constructions".  I know it will work and I was just noting that the inheritance might not make sense for all Abridged versions.  We can go ahead and use your proposed inheritance plan for this Construction case, though, and we will worry about how the other non-Abridged versions will work later.

**Comment by mostaphaRoudsari** [comment]
As I added above I think we should use inheritance when it works and if there is a case that doesn't work that way they you don't use it not the other way around.

**Pull request merged**